modules/javafx.graphics/src/main/java/com/sun/javafx/scene/layout/SpaceDistributor.java

```diff
@@ -0,0 +1,258 @@
+
+            for (int i = 0; i < count; i++) {
+                double size = sizes[i];
+                double targetSize = targetSizes[i];
+
+                if ((grow && size < targetSize) || (!grow && size > targetSize)) {
+                    double x = round(size + idealChange, renderScale);
+                    double newSpace = grow ? Math.min(targetSize, x) : Math.max(targetSize, x);
+
+                    available = round(available - newSpace + size, renderScale);
+                    sizes[i] = newSpace;
+
+                    if (newSpace == targetSize && --resizables == 0) {  // exit method when nothing left to resize
+                        return;
+                    }
+                }
+            }
+
+            if (initialResizables == resizables) {  // exit loop when nothing changed
+                break;
+            }
+        }
+
+        if (renderScale == 0) {  // exit after distribution when not dealing with pixels
+            return;
+        }
+
+        /*
+         * When dealing with pixel sized chunks, a final pass is needed to distribute left over
+         * space that could not be distributed evenly to all spaces that did not reach their
+         * target, meaning some spaces will get or lose one extra pixel.
+         */
+
+        double pixelSize = 1 / renderScale;
+
+        //assert Math.abs(available) <= round(pixelSize * resizables, renderScale) : "available exceeded threshold: " + available + "; resizables: = " + resizables;
+
+        for (int i = 0; i < count; i++) {
+            if (Math.abs(available) < 0.5 * pixelSize) {  // exit if less than half a pixel is available
+                break;
+            }
+
+            if (available < 0) {
+                double s = round(sizes[i] - pixelSize, renderScale);
+
+                if (s >= targetSizes[i]) {
+                    sizes[i] = s;
+                    available += pixelSize;
+                }
+            }
+            else {
+                double s = round(sizes[i] + pixelSize, renderScale);
+
+                if (s <= targetSizes[i]) {
+                    sizes[i] = s;
+                    available -= pixelSize;
+                }
+            }
+        }
+    }
+
+    // array functions:
+
+    private static double[] ceil(double[] values, double s) {
+        double[] ceiledValues = new double[values.length];
+
+        for (int i = 0; i < values.length; i++) {
+            ceiledValues[i] = ceil(values[i], s);
+        }
+
+        return ceiledValues;
+    }
+
+    private static double sum(double[] values) {
+        double sum = 0;
+
+        for (int i = 0; i < values.length; i++) {
+            sum += values[i];
+        }
+
+        return sum;
+    }
+
+    // rounding functions:
+
+    private static double down(double x, double s) {
+        return x < 0 ? ceil(x, s) : floor(x, s);
+    }
+
+    private static double ceil(double x, double s) {
+        if (s == 0) {
+            return x;
+        }
```

Comment:
if we decouple 'deal with pixels' and scale, the array functions will be simpler, and we would not need to do == 0 every time (use a boolean flag?)
Also, may be spell 'scale' for clarity?